src/ispyb/cli/job.py

```diff
@@ -0,0 +1,471 @@
+
+    if zocalo:
+        trigger_variables = []
+        for p in options.triggervariables:
+            if ":" not in p:
+                sys.exit("Invalid trigger variable specification: " + p)
+            trigger_variables.append(p.split(":", 1))
+
+    jp = i.mx_processing.get_job_params()
+    # _job_params = StrictOrderedDict([('id', None), ('datacollectionid', None), ('display_name', None), ('comments', None), ('recipe', None), ('automatic', None)])
+    jp["automatic"] = options.source == "automatic"
+    jp["comments"] = options.comment
+    jp["datacollectionid"] = dcid or sweeps[0][0]
+    jp["display_name"] = options.display
+    jp["recipe"] = options.recipe
+    print("Creating database entries...")
+
+    jobid = i.mx_processing.upsert_job(list(jp.values()))
+    print(f"  JobID={jobid}")
+    for key, value in parameters:
+        jpp = i.mx_processing.get_job_parameter_params()
+        # _job_parameter_params = StrictOrderedDict([('id', None), ('job_id', None), ('parameter_key', None), ('parameter_value', None)])
+        jpp["job_id"] = jobid
+        jpp["parameter_key"] = key
+        jpp["parameter_value"] = value
+        jppid = i.mx_processing.upsert_job_parameter(list(jpp.values()))
+        print(f"  JPP={jppid}")
+
+    for sweep in sweeps:
+        jisp = i.mx_processing.get_job_image_sweep_params()
+        # _job_image_sweep_params = StrictOrderedDict([('id', None), ('job_id', None), ('datacollectionid', None), ('start_image', None), ('end_image', None)])
+        jisp["job_id"] = jobid
+        jisp["datacollectionid"] = sweep[0]
+        jisp["start_image"] = sweep[1]
+        jisp["end_image"] = sweep[2]
+        jispid = i.mx_processing.upsert_job_image_sweep(list(jisp.values()))
+        print(f"  JISP={jispid}")
+
+    print(f"All done. Processing job {jobid} created")
+    print()
+    if zocalo:
+        if options.trigger:
+            go_call = ["zocalo.go", "-p", str(jobid)]
+            for kv in trigger_variables:
+                go_call.append("--set=%s=%s" % (kv[0], kv[1]))
+            result = procrunner.run(go_call)
+            if result["exitcode"] or result["stderr"]:
+                sys.exit("Error triggering processing job")
+            print("Successfully triggered processing job")
+            print()
+
+        else:
+            print("To trigger the processing job you now need to run:")
+            print(f"  zocalo.go -p {jobid}")
+            print()
+
+    return jobid
+
+
+def run():
+    parser = OptionParser(
+        usage="ispyb.job [options] JOBID",
+        description="Command line tool to manipulate ISPyB processing table entries.",
+    )
+
+    if os.path.isdir("/dls_sw/apps/zocalo/live/recipes"):
+        available_recipes = filter(
+            lambda r: r.startswith("ispyb-") and r.endswith(".json"),
+            os.listdir("/dls_sw/apps/zocalo/live/recipes"),
+        )
+        available_recipes = sorted(r[6:-5] for r in available_recipes)
```

Comment:
Could the location of this file be provided via a command line parameter (or alternatively an environment variable)? This _could_ default to the `/dls_sw/` location, but I'm open to arguments against this. See e.g. the `--credentials` parameter to `ispyb.last_data_collections_on` (which will now default to reading from the `ISPYB_CREDENTIALS` environment variable if not explicitly provided on the command line):
https://github.com/DiamondLightSource/ispyb-api/blob/7e9f504f6765a7070a4680e4c2a286510f5da6db/src/ispyb/cli/last_data_collections_on.py#L128

Reply:
Ideally this sort of stuff would be resolved using a zocalo configuration yaml file that not only sets a recipe location, but also stuff like graylog logging, access to activemq/rabbitmq credentials, etc.

I suppose temporary gorilla variables are fine, but I would only like to add them only if someone other than us is _actually_ using them

Reply:
I did however take the opportunity to clear up this idiosyncratic filter lambda generator :slightly_smiling_face: